### ISSUE TITLE
[bug] Draw dest is shifted for next draw

### ISSUE THREAD

**Issue by snsvrno** [ggez/ggez#462]
# Background

On a simple project, with 1 `spritebatch` and 1 `text` object. Working with resize code to make `spritebatch` center while keeping the text in the top left corner.

The sprites in the `spritebatch` are drawn with 0,0 being the top right, then the `spritebatch` itself is shifted over using a `Point2` while drawing it to the screen/context. 

Using 
- **ggez** 0.4.3
- **rust** 1.28.0 / stable x86_64-pc-windows-msvc

# Code

```rust
fn draw(&mut self, context : &mut ggez::Context) -> GameResult<()> {
        ggez::graphics::clear(context);

        // draws the board
        ggez::graphics::draw(context,&self.sprite_batch,self.board_draw_pos,0.0)?;

        // draws UI text
        ggez::graphics::draw(context,&self.move_count_obj,ggez::graphics::Point2::new(10.0,10.0),0.0)?;

        ggez::graphics::present(context);

        Ok(())
    }
```

**Issue**:

When I draw the `spritebatch` first, and then the text, the text seems to assume the "0,0" has shifted based on the previous draw. So the text is effectively drawn at `0 + self.board_draw_pos.obj.x, 0.0 + self.board_draw_pos.obj.y`.

After some testing I found that it only happens to the draw immediately afterwards.

## Case 1

```rust
        ggez::graphics::draw(context,&self.sprite_batch,self.board_draw_pos,0.0)?;
        ggez::graphics::draw(context,&self.text,ggez::graphics::Point2::new(10.0,10.0),0.0)?;
```
`spritebatch` is drawn at `self.board_draw_pos.obj.x, self.board_draw_pos.obj.y`
`text` is drawn at `self.board_draw_pos.obj.x + 10.0, self.board_draw_pos.obj.y + 10.0`

I can see some "blinking" during the redraws, like it draws in `10.0, 10.0` for one or two frames and then jumps back to the `self.board_draw_pos.obj.x + 10.0, self.board_draw_pos.obj.y + 10.0` position.

![image](https://user-images.githubusercontent.com/36494927/44235232-09537280-a166-11e8-8845-8e03815a60e6.png)

Fails .

## Case 2

```rust
        ggez::graphics::draw(context,&self.text,ggez::graphics::Point2::new(10.0,10.0),0.0)?;
        ggez::graphics::draw(context,&self.sprite_batch,self.board_draw_pos,0.0)?;
```
`spritebatch` is drawn at `self.board_draw_pos.obj.x, self.board_draw_pos.obj.y`
`text` is drawn at `10.0, 10.0`

![image](https://user-images.githubusercontent.com/36494927/44235256-1ff9c980-a166-11e8-90e9-dd3331b15983.png)

Passes .

## Case 3

```rust
        ggez::graphics::draw(context,&self.sprite_batch,self.board_draw_pos,0.0)?;
        ggez::graphics::draw(context,&self.text,ggez::graphics::Point2::new(10.0,10.0),0.0)?;
        ggez::graphics::draw(context,&self.text,ggez::graphics::Point2::new(10.0,10.0),0.0)?;
```
`spritebatch` is drawn at `self.board_draw_pos.obj.x, self.board_draw_pos.obj.y`
`text` is drawn at both `10.0, 10.0` and  `self.board_draw_pos.obj.x + 10.0, self.board_draw_pos.obj.y + 10.0`

![image](https://user-images.githubusercontent.com/36494927/44235298-3b64d480-a166-11e8-990b-6d089c06b3cb.png)

Passes ? ish

**Comment by icefoxen** [comment]
...well that's interesting.

**Comment by icefoxen** [comment]
Okay, this is *really weird* behavior, since there is (in theory) no possible way that the `draw()` call for the `Text` knows anything at all about the location of the `SpriteBatch`.  This makes me wonder if you're doing something strange with your coordinate systems or such.

Also I can't reproduce this error, unfortunately.  Is it possible to share your full code so we can dig into this in more detail?  My own example program is listed in full below, can you try it out and see if you get the correct or incorrect behavior?

```rust
extern crate ggez;

use ggez::conf;
use ggez::event;
use ggez::graphics;
use ggez::{Context, GameResult};
use std::env;
use std::path;

struct MainState {
    text: graphics::Text,
    batch: graphics::spritebatch::SpriteBatch,

}

impl MainState {
    fn new(ctx: &mut Context) -> GameResult<MainState> {
        let font = graphics::Font::new(ctx, "/DejaVuSerif.ttf", 48)?;
        let text = graphics::Text::new(ctx, "Hello world!", &font)?;

        let image = graphics::Image::new(ctx, "/player.png")?;
        let batch = graphics::spritebatch::SpriteBatch::new(image);
        let s = MainState { text, batch };
        Ok(s)
    }
}

impl event::EventHandler for MainState {
    fn update(&mut self, _ctx: &mut Context) -> GameResult<()> {
        Ok(())
    }

    fn draw(&mut self, ctx: &mut Context) -> GameResult<()> {
        graphics::clear(ctx);

        for i in 0..10 {
            let fi = i as f32 * 20.0;
            self.batch.add(graphics::DrawParam {
                dest: graphics::Point2::new(fi, fi),
                .. graphics::DrawParam::default()
            });
        }

        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 300.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(10.0, 10.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(10.0, 10.0), 0.0)?;
        graphics::present(ctx);

        Ok(())
    }
}

pub fn main() {
    let c = conf::Conf::new();
    let ctx = &mut Context::load_from_conf("heckin_strange", "ggez", c).unwrap();
    if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
        let mut path = path::PathBuf::from(manifest_dir);
        path.push("resources");
        ctx.filesystem.mount(&path, true);
    }

    let state = &mut MainState::new(ctx).unwrap();
    if let Err(e) = event::run(ctx, state) {
        println!("Error encountered: {}", e);
    } else {
        println!("Game exited cleanly.");
    }
}
```

What I get:

![image](https://user-images.githubusercontent.com/1335133/44476697-92efbe00-a606-11e8-84d9-98335a3a3c12.png)

**Comment by snsvrno** [comment]
Soooo..... It doesn't work for me.

![image](https://user-images.githubusercontent.com/36494927/44485687-a0a94180-a60e-11e8-805a-cf7d9fd63846.png)

Exact same code as yours (copy and paste). I do use a build.rs script for SDL though, taken from [the SDL-RS repo instructions](https://github.com/Rust-SDL2/rust-sdl2#windows-with-build-script). Using MSVC on Windows 10 with SDL 2.0.8, x64.

So looks like an SDL issue? I don't have another windows instance to test it on, but based on your sample it looks like there is something wrong locally ...

**Comment by snsvrno** [comment]
To help narrow it out, it looks like a TEXT issue and not a drawing issue. I mixed up the order a bit and it didn't "bug" as expected.

![image](https://user-images.githubusercontent.com/36494927/44486175-12ce5600-a610-11e8-9a78-24f40c255bdc.png)

Same code but moved around the drawing and changed up the Y on the text.

```rust
        // drawing inside of MainState::Draw
        graphics::draw(ctx, &self.text, graphics::Point2::new(40.0, 40.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 300.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(40.0, 40.0), 0.0)?;
```

Based on my understanding of the issue I'd expect the spritebatch to be moved to `40,340` but that doesn't happen. So it looks it is somehow isolated to the text drawing.

**Comment by icefoxen** [comment]
Text is drawn basically the same as anything else; a `Text` object is just an `Image` and some metadata.  I'll see if I can reproduce this on windows.

**Comment by snsvrno** [comment]
More, so I don't think its an text thing anymore. Did some more tests and it looks like the `spritebatch` is somehow causing it.

```rust
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
```
This draws as expected

![image](https://user-images.githubusercontent.com/36494927/44486760-b8ce9000-a611-11e8-8488-047a81242b43.png)

---

But whenever I move a text or image after the batch it inherits the position...

![image](https://user-images.githubusercontent.com/36494927/44486906-1c58bd80-a612-11e8-8e0b-8ab912ea6b0f.png)

```rust
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
```

![image](https://user-images.githubusercontent.com/36494927/44486899-17940980-a612-11e8-9ea8-582d8e548e14.png)

```rust
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.image, graphics::Point2::new(0.0, 150.0), 0.0)?;
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 150.0), 0.0)?;
```

**Comment by icefoxen** [comment]
okay...  So let me think about this. ggez has one instance buffer that stores drawing information, ie the inputs to the vertex shader, `graphics::InstanceProperties`.  If you draw an image, it fills that instance buffer with one `InstanceProperties` and issues a draw command.  If you draw a SpriteBatch, it fills that instance buffer with one `InstanceProperties` per item you've added to the batch and issues a draw command.

The only way I can see this could be happening is basically that the instance buffer is being corrupted somehow, such as an `InstanceProperties` being saved to the wrong index or something like that.  But this is all just `gfx-rs`/OpenGL code, so its effects *shouldn't* be platform-dependent and shouldn't depend on SDL versions unless there's a driver bug involved.

**Comment by snsvrno** [comment]
More fuel for you. I decided to clone GGEZ and play with `draw_ex` to figure out what is going on behind the scenes. I didn't find anything interesting, but I noticed there was code that was converting `dest` into transformations and applying them (so essentially apply a transformation and drawing at `0,0` is the same as using a transformation of `0,0` and `dest`).

So I decided to `pop` right after the batch call. Results ....

## Code Set 1

```rust
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 50.0), 0.0)?;
        //graphics::pop_transform(ctx);
        //graphics::apply_transformations(ctx);
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 0.0), 0.0)?;
        graphics::present(ctx);
```

![image](https://user-images.githubusercontent.com/36494927/44488689-9ee37c00-a616-11e8-9309-03bc4dfd1c83.png)

## Code Set 2
```rust
        graphics::draw(ctx, &self.batch, graphics::Point2::new(0.0, 50.0), 0.0)?;
        graphics::pop_transform(ctx);
        graphics::apply_transformations(ctx);
        graphics::draw(ctx, &self.text, graphics::Point2::new(0.0, 0.0), 0.0)?;
        graphics::present(ctx);
```

![image](https://user-images.githubusercontent.com/36494927/44488752-c9353980-a616-11e8-837b-08ef63b75f6e.png)

---
I don't call any transforms, so I'd thing it wouldn't do anything, but it fixes it...

I tried to use `println!("Transform: {:?}",graphics::get_transform(ctx));` throughout to see if I can trace a rogue transform, but they all have position `0,0`

**Comment by snsvrno** [comment]
So I tried looking with RenderDoc, and I think its an issue with SDL / Driver issue. First of all I have no idea what I'm doing, so I'll attach the capture [drawing_offset.zip (drawing_offset.rdc)](https://github.com/ggez/ggez/files/2314570/drawing_offset.zip).

When I load up the capture it draws wrong.

![image](https://user-images.githubusercontent.com/36494927/44527988-a69e3180-a6a5-11e8-8107-5d3bf6d78f29.png)

But then when I play with the overlay while selecting the text draw it snaps to the right spot ... (first _highlight_ and then _none_)
![image](https://user-images.githubusercontent.com/36494927/44528025-c170a600-a6a5-11e8-99e9-bc4b74f03f31.png)
![image](https://user-images.githubusercontent.com/36494927/44528039-ca617780-a6a5-11e8-96f1-86545351e8e0.png)

But then if I click anything else in the _event browser_ snaps back to the wrong spot.

Also, this shows me its trying to do the right thing but failing somehow.
![image](https://user-images.githubusercontent.com/36494927/44528154-114f6d00-a6a6-11e8-80fb-1b3085c9e03e.png)

**Comment by icefoxen** [comment]
This was a bug in transform stuff, should be fixed as of 0.6's transform stack rework
